Fail to read .exe in .7z

When extracting a file containing an `exe` file, 
the different reading methods for the reader will give different results and generate panic.

Use `io.Copy` will be fine.

```golang
	reader, _ := sevenzip.OpenReader(fp)
	for _, fileIn7z := range reader.File {
		r, _ := fileIn7z.Open()
		var buffer bytes.Buffer
		io.Copy(&buffer, r)
		defer reader.Close()
		data := buffer.Bytes()

		fmt.Println("Data size:", len(data))
		// use data to do something...
	}
```

Use `ioutil.ReadAll`  will be ok in most cases.
But if the coming file is `.exe`, it will cause a panic: `panic: runtime error: slice bounds out of range [:5] with capacity 3`

```golang
	reader, _ := sevenzip.OpenReader(fp)
	for _, fileIn7z := range reader.File {
		r, _ := fileIn7z.Open()
		data, _ := ioutil.ReadAll(r)

		fmt.Println("Data size:", len(data))
		// use data to do something...
	}
```

Here is the complete test code.

```golang
package main

import (
	"bytes"
	"fmt"
	"io"
	"io/ioutil"
	"path"

	"github.com/bodgit/sevenzip"
)

func checkErr(err error) {
	if err != nil {
		panic(err)
	}
}

func readByIoutil(fp string) {
	fmt.Println("[readByIoutil] start")

	reader, err := sevenzip.OpenReader(fp)
	checkErr(err)
	for _, fileIn7z := range reader.File {
		filename := path.Base(fileIn7z.Name)
		fmt.Println("Processing", filename)
		reader, err := fileIn7z.Open()
		checkErr(err)
		data, err := ioutil.ReadAll(reader)
		checkErr(err)

		fmt.Println("Data size:", len(data))
		// use data to do something...
	}
	fmt.Println("[readByIoutil] finished\n")
}

func readByCopy(fp string) {
	fmt.Println("[readByCopy] start")
	reader, err := sevenzip.OpenReader(fp)
	checkErr(err)
	for _, fileIn7z := range reader.File {
		filename := path.Base(fileIn7z.Name)
		fmt.Println("Processing", filename)
		r, err := fileIn7z.Open()
		checkErr(err)
		var buffer bytes.Buffer
		_, err = io.Copy(&buffer, r)
		defer reader.Close()
		data := buffer.Bytes()

		fmt.Println("Data size:", len(data))
		// use data to do something...
	}
	fmt.Println("[readByCopy] finished\n")
}

func main() {
	fp := "putty.7z"

	readByCopy(fp)
	readByIoutil(fp)
}

```

And may get output like below

```
[readByCopy] start
Processing putty.exe
Data size: 1647912
[readByCopy] finished

[readByIoutil] start
Processing putty.exe
panic: runtime error: slice bounds out of range [:5] with capacity 3

goroutine 1 [running]:
```


I used `putty.exe` and compressed it into `putty.7z` as a test file, which is also attached here.
(`putty` is a common `ssh` tool used by windows user)

Source:  https://www.chiark.greenend.org.uk/~sgtatham/putty/latest.html
64-bit x86: [putty.exe](https://the.earth.li/~sgtatham/putty/latest/w64/putty.exe)

File information:

```bash
$ file putty.exe 
putty.exe: PE32+ executable (GUI) x86-64, for MS Windows
```

```bash
$ 7z a putty.7z putty.exe 

7-Zip [64] 17.05 : Copyright (c) 1999-2021 Igor Pavlov : 2017-08-28
p7zip Version 17.05 (locale=utf8,Utf16=on,HugeFiles=on,64 bits,8 CPUs LE)

Open archive: putty.7z
--
Path = putty.7z
Type = 7z
Physical Size = 816140
Headers Size = 122
Method = LZMA2:21 BCJ
Solid = -
Blocks = 1

Scanning the drive:
1 file, 1647912 bytes (1610 KiB)

Updating archive: putty.7z

Items to compress: 1

     
Files read from disk: 1
Archive size: 816140 bytes (798 KiB)
```

```
$ file putty.7z
putty.7z: 7-zip archive data, version 0.4
```

Attached File:
[putty.7z.zip](https://github.com/bodgit/sevenzip/files/11765815/putty.7z.zip)

--- a/internal/bra/bra.go
+++ b/internal/bra/bra.go
@@ -12,3 +12,11 @@ func max(x, y int) int {
 
 	return y
 }
+
+func min(x, y int) int {
+	if x < y {
+		return x
+	}
+
+	return y
+}

--- a/internal/bra/reader.go
+++ b/internal/bra/reader.go
@@ -9,6 +9,7 @@ import (
 type readCloser struct {
 	rc   io.ReadCloser
 	buf  bytes.Buffer
+	n    int
 	conv converter
 }
 
@@ -30,13 +31,19 @@ func (rc *readCloser) Read(p []byte) (int, error) {
 		if !errors.Is(err, io.EOF) {
 			return 0, err
 		}
+
+		if rc.buf.Len() < rc.conv.Size() {
+			rc.n = rc.buf.Len()
+		}
 	}
 
-	if n := rc.conv.Convert(rc.buf.Bytes(), false); n > 0 {
-		return rc.buf.Read(p[:min(n, len(p))])
-	}
+	rc.n += rc.conv.Convert(rc.buf.Bytes()[rc.n:], false)
 
-	return rc.buf.Read(p)
+	n, err := rc.buf.Read(p[:min(rc.n, len(p))])
+
+	rc.n -= n
+
+	return n, err
 }
 
 func newReader(readers []io.ReadCloser, conv converter) (io.ReadCloser, error) {

--- a/internal/bra/reader.go
+++ b/internal/bra/reader.go
@@ -33,7 +33,7 @@ func (rc *readCloser) Read(p []byte) (int, error) {
 	}
 
 	if n := rc.conv.Convert(rc.buf.Bytes(), false); n > 0 {
-		return rc.buf.Read(p[:n])
+		return rc.buf.Read(p[:min(n, len(p))])
 	}
 
 	return rc.buf.Read(p)

--- a/reader_test.go
+++ b/reader_test.go
@@ -333,3 +333,19 @@ func BenchmarkBrotli(b *testing.B) {
 func BenchmarkZstandard(b *testing.B) {
 	benchmarkArchive(b, "zstd.7z")
 }
+
+func BenchmarkBCJ(b *testing.B) {
+	benchmarkArchive(b, "bcj.7z")
+}
+
+func BenchmarkPPC(b *testing.B) {
+	benchmarkArchive(b, "ppc.7z")
+}
+
+func BenchmarkARM(b *testing.B) {
+	benchmarkArchive(b, "arm.7z")
+}
+
+func BenchmarkSPARC(b *testing.B) {
+	benchmarkArchive(b, "sparc.7z")
+}

--- a/reader_test.go
+++ b/reader_test.go
@@ -8,6 +8,7 @@ import (
 	"path/filepath"
 	"testing"
 	"testing/fstest"
+	"testing/iotest"
 
 	"github.com/bodgit/sevenzip"
 	"github.com/bodgit/sevenzip/internal/util"
@@ -28,7 +29,7 @@ func readArchive(t *testing.T, r *sevenzip.ReadCloser) {
 
 		h.Reset()
 
-		if _, err := io.Copy(h, rc); err != nil {
+		if _, err := io.Copy(h, iotest.OneByteReader(rc)); err != nil {
 			t.Fatal(err)
 		}
 

